overlord/snapstate/snapstate.go

```diff
@@ -1407,13 +1407,85 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, revOpts []
 	return updateManyFiltered(ctx, st, names, revOpts, userID, nil, flags, "")
 }
 
-// EnforceSnaps installs/updates/removes snaps reported by validationErrorToSolve.
-// validationSets is the list of sets passed by the user and it's used in the
-// final stage to update validation-sets tracking in the state.
-// userID is used for store auth.
-func EnforceSnaps(ctx context.Context, st *state.State, validationSets []string, validationErrorToSolve *snapasserts.ValidationSetsValidationError, userID int) (tasksets []*state.TaskSet, names []string, err error) {
-	// TODO
-	return nil, nil, fmt.Errorf("not implemented")
+// ResolveValidationSetsEnforcementError installs and updates snaps reported by validationErrorToSolve.
+func ResolveValidationSetsEnforcementError(ctx context.Context, st *state.State, valErr *snapasserts.ValidationSetsValidationError, pinnedSeqs map[string]int, userID int) ([]*state.TaskSet, []string, error) {
+	if len(valErr.InvalidSnaps) != 0 {
+		invSnaps := make([]string, 0, len(valErr.InvalidSnaps))
+		for invSnap := range valErr.InvalidSnaps {
+			invSnaps = append(invSnaps, invSnap)
+		}
+		return nil, nil, fmt.Errorf("cannot auto-resolve enforcement constraints that require removing snaps: %s", strutil.Quoted(invSnaps))
+	}
+
+	affected := make([]string, 0, len(valErr.MissingSnaps)+len(valErr.WrongRevisionSnaps))
+	var tasksets []*state.TaskSet
+	// use the same lane for installing and refreshing so everything is reversed
+	lane := st.NewLane()
+
+	collectRevOpts := func(snapToRevToVss map[string]map[snap.Revision][]string) ([]string, []*RevisionOptions) {
+		var names []string
+		var revOpts []*RevisionOptions
+
+		for snapName, revAndVs := range snapToRevToVss {
+			for rev, valsets := range revAndVs {
+				vsKeys := make([]snapasserts.ValidationSetKey, 0, len(valsets))
+				for _, vs := range valsets {
+					vsKey := snapasserts.NewValidationSetKey(valErr.Sets[vs])
+					vsKeys = append(vsKeys, vsKey)
+				}
+
+				revOpts = append(revOpts, &RevisionOptions{Revision: rev, ValidationSets: vsKeys})
+			}
+			names = append(names, snapName)
+		}
+
+		return names, revOpts
+	}
+
+	if len(valErr.WrongRevisionSnaps) > 0 {
+		names, revOpts := collectRevOpts(valErr.WrongRevisionSnaps)
+		flags := &Flags{Transaction: client.TransactionAllSnaps, Lane: lane, NoReRefresh: true}
+
+		updated, tss, err := UpdateMany(ctx, st, names, revOpts, userID, flags)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot auto-resolve enforcement constraints: %v", err)
```

Comment:
same, s/enforcement/validation/ ? or enforced validation

Reply:
we probably don't want to wrap ChangeConflictErrors though, unless we do the work to make the check for those use the modern error wrapping support